{storage, memory, examples}: mysql implementation for memory service

Rememorio opened the pull request:
- Add memory implementation for mysql.
- Add examples and docs.
- Ref: https://github.com/trpc-group/trpc-agent-go/pull/571

Rememorio committed:
storage/mysql: init

Rememorio committed:
memory/mysql: init

Rememorio committed:
examples/memory: add mysql

Rememorio committed:
docs/memory: add mysql

Rememorio committed:
examples: remove toolchain and update go.sum dependencies

Rememorio committed:
Merge branch 'main' into memory_mysql

Rememorio committed:
chore: update trpc-a2a-go dependency to v0.2.5

- Updated the version of trpc.group/trpc-go/trpc-a2a-go from v0.2.4 to v0.2.5 in go.mod and go.sum files to incorporate the latest changes and improvements.

Rememorio committed:
refactor: update database client handling to use ClientInterface

- Changed the mockClientBuilder and related functions to return ClientInterface instead of *sql.DB, enhancing abstraction for testing.
- Updated service and test files to reflect the new ClientInterface type, improving testability and flexibility in database operations.
- Added new tests to ensure compliance with the ClientInterface, validating that the interface is correctly implemented.

Rememorio committed:
test: enhance memory service tests for error handling and edge cases

- Added new tests for AddMemory, UpdateMemory, and ReadMemories functions to cover scenarios with no memory limit, select errors, unmarshal errors, and update errors.
- Improved error handling tests for SearchMemories to validate behavior when scanning rows fails and when unmarshaling invalid JSON.
- Ensured comprehensive coverage of edge cases and error scenarios, enhancing the robustness of the memory service tests.

Rememorio committed:
Merge branch 'main' into memory_mysql

Rememorio committed:
Merge branch 'main' into memory_mysql

Rememorio committed:
chore: update trpc-a2a-go dependency and refactor client builder function names

- Updated trpc-a2a-go dependency version in go.mod and go.sum.
- Refactored function names from DefaultClientBuilder to defaultClientBuilder for consistency in mysql package.
- Adjusted related test cases to reflect the updated function name.

Rememorio committed:
refactor: update MySQL service to use new Client interface and improve error handling

- Changed the MySQL service to utilize the new Client interface, replacing ClientInterface with Client.
- Updated method signatures and implementations to align with the new interface, including Exec, Query, and QueryRow methods.
- Enhanced error messages for better clarity during database operations.
- Added new tests for sorting memories and handling edge cases in memory searches.
- Refactored existing tests to accommodate the changes in the Client interface.

Rememorio committed:
feat(mysql): add MySQL client implementation and testing

- Introduced MySQL client interface with methods for executing queries, transactions, and connection management.
- Implemented a default client builder with configurable options for connection pooling and DSN.
- Added comprehensive unit tests to validate client functionality and behavior.
- Created go.mod and go.sum files for dependency management.

Rememorio committed:
refactor(mysql): rename SQLDBClient to sqlDBClient and update field access

- Renamed SQLDBClient struct to sqlDBClient for consistency with Go naming conventions.
- Updated field access from public to private by changing DB to db.
- Adjusted all related test cases to reflect the new struct name and field access.

Rememorio committed:
Merge branch 'storage_mysql' into memory_mysql

Rememorio committed:
refactor(mysql): remove NewClient function from sqlDBClient

- Removed the NewClient function that created a new Client from an existing *sql.DB connection.
- This change simplifies the client creation process and aligns with the current design approach.

Rememorio committed:
Merge branch 'storage_mysql' into memory_mysql

Rememorio committed:
feat(mysql): add WrapSQLDB function for internal client wrapping

- Introduced WrapSQLDB function to wrap a *sql.DB connection into a Client.
- Added documentation to clarify that this function is for internal use only and should not be called directly from external packages.

Rememorio committed:
Merge branch 'storage_mysql' into memory_mysql

Rememorio committed:
refactor(mysql): update service tests to use WrapSQLDB

- Replaced direct instantiation of SQLDBClient with WrapSQLDB in service tests.
- Ensured consistency across test cases by utilizing the new wrapping function for database connections.

Rememorio committed:
Merge branch 'main' into memory_mysql

Rememorio committed:
memory/mysql: improve table name validation error message

- Introduced a constant for maximum table name length to enhance readability.
- Updated error message to include the maximum length in the validation check.

Rememorio committed:
memory/mysql: enhance table name validation in service options

- Added validation for table names in the WithTableName function to prevent SQL injection.
- Removed redundant table name validation from NewService function and updated comments for clarity.

Rememorio committed:
Merge branch 'main' into memory_mysql

hyprh commented on memory/mysql/service.go:
```diff
@@ -0,0 +1,431 @@
+			memory_id VARCHAR(64) NOT NULL,
+			memory_data JSON NOT NULL,
+			created_at TIMESTAMP NOT NULL DEFAULT CURRENT_TIMESTAMP,
+			updated_at TIMESTAMP NOT NULL DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP,
```
default current_timestamp on update current_timestamp 这个用法在老版本的mysql可能会有坑，但是现在这么用估计也OK

Rememorio committed:
memory/mysql: implement soft delete functionality for memory records

- Updated DeleteMemory and ClearMemories methods to perform soft deletes by setting a deleted_at timestamp instead of removing records.
- Modified relevant test cases to reflect changes in deletion logic.
- Adjusted queries in ReadMemories and other methods to exclude soft-deleted records.

Rememorio committed:
memory/mysql: add soft delete option and update related functionality

- Introduced `WithSoftDelete` option to enable soft delete behavior for memory records.
- Updated memory service methods to handle soft deletes, including adjustments to delete, read, and search queries.
- Enhanced documentation in both English and Chinese to reflect the new soft delete feature.

Rememorio committed:
memory/mysql: update table creation behavior and documentation

- Removed the `WithAutoCreateTable` option as table creation is now always performed during service initialization.
- Enhanced documentation to clarify that the table is automatically created and will panic if creation fails.
- Updated validation for table names to ensure they meet specific criteria, with panics for invalid names.
- Adjusted related tests to reflect the new behavior regarding table creation and validation.

hyprh commented on memory/mysql/service.go:
```diff
@@ -0,0 +1,440 @@
+		return fmt.Errorf("unmarshal memory entry failed: %w", err)
+	}
+
+	now := time.Now()
```
这里用time.now.utc()可能好一点，mysql timestamp没有存时区

Rememorio replied:
fixed，这里的都改了

Rememorio committed:
memory/mysql: use utc instead of local zone

hyprh reviewed (approved):
LGTM
